### PR TITLE
Refactor/bucket object in state

### PR DESCRIPTION
**Pull request by shino**
Some independent logics fetch bucket object from Riak, bucket policy, bucket ACL
and bucket owner retrieval. In object access, e.g. GET Object, we can/should suppress
extra bucket object fetch.

Refactoring plan is simple. Let object-related webmachine callback
resources keep bucket objects (riak objects) in their state.

This PR is motivated mainly by https://github.com/basho/riak_cs/issues/701 .
Also this will be the basis for https://github.com/basho/riak_cs/issues/723 .

In testing this PR, two things found, both are reproduced at current develop branch.
1. Dialyzer outputs some warnings:
2. make test fails at `riak_cs_get_fsm_eqc`
These may/should be treated separated issues.

1 Dialyzer

```
riak_cs_gc.erl:187: The pattern [{UUID, _} | _] can never match the type []
riak_cs_gc_d.erl:173: The call riak_cs_gc_d:fetch_eligible_manifest_keys(RiakPid::'undefined' | pid(),BatchStart::non_neg_integer(),'undefined') will never return since it differs in the 3rd argument from the success typing arguments: (pid(),non_neg_integer(),binary())
riak_cs_gc_d.erl:608: Record construction #state{interval::'undefined' | non_neg_integer(),last::'undefined' | non_neg_integer(),next::{{non_neg_integer(),1 | 2 | 3 | 4 | 5 | 6 | 7 | 8 | 9 | 10 | 11 | 12,1..255},{byte(),byte(),byte()}},current_files::'undefined' | [#lfs_manifest_v3{version::integer(),block_size::'undefined' | integer(),bkey::'undefined' | {binary(),binary()},metadata::'undefined' | [{_,_}],uuid::'undefined' | binary(),content_length::'undefined' | non_neg_integer(),content_type::'undefined' | binary(),state::'active' | 'deleted' | 'pending_delete' | 'scheduled_delete' | 'undefined' | 'writing',write_blocks_remaining::'undefined' | [integer()],delete_blocks_remaining::'undefined' | [integer()],acl::'no_acl_yet' | 'undefined' | #acl_v1{owner::{string(),string()} | {string(),string(),string()},grants::[{'AllUsers' | 'AuthUsers' | {string(),string()},['FULL_CONTROL' | 'READ' | 'READ_ACP' | 'WRITE' | 'WRITE_ACP']}],creation_time::{non_neg_integer(),non_neg_integer(),non_neg_integer()}} | #acl_v2{owner::{string(),string()} | {string(),string(),string()},grants::[{'AllUsers' | 'AuthUsers' | {string(),string()},['FULL_CONTROL' | 'READ' | 'READ_ACP' | 'WRITE' | 'WRITE_ACP']}],creation_time::{non_neg_integer(),non_neg_integer(),non_neg_integer()}},props::'undefined' | [atom() | tuple()],cluster_id::'undefined' | binary()}],current_fileset::'undefined' | {set(),set()},current_riak_object::'undefined' | {'riakc_obj','undefined' | binary(),'undefined' | binary(),'undefined' | binary(),'undefined' | [{dict(),binary()}],'undefined' | dict(),'undefined' | binary()},batch_count::non_neg_integer(),batch_skips::non_neg_integer(),batch::'undefined' | [binary()],manif_count::non_neg_integer(),block_count::non_neg_integer(),pause_state::atom(),interval_remaining::'undefined' | non_neg_integer(),timer_ref::reference(),delete_fsm_pid::'undefined' | pid(),continuation::'undefined' | binary()} violates the declared type of field next::'undefined' | non_neg_integer()
Unknown functions:
  riak_object:bucket/1
Unknown types:
  riakc_pb_socket:index_results/0
 done in 0m28.78s
done (warnings were emitted)
```

2 make test (sudden death, almost no output but failed)

```
$ make test
(SNIP)
.......................
[test] Killed
```

### CODE REVIEW

**Comment by kuenishi** [comment]
I succeeded 5/5 times on `make test`. Probably not enough memory? I failed at eqc several times like yours when I had only 5-6 gigs RAM. At that time restarting the OS worked for me, probably because of memory fragmentation, I guess.

**Comment by shino** [comment]
@kuenishi Thanks! There were OOM Killer logs...

**Comment by kuenishi** [comment]
- [x] review
- [x] riak_test
- [x] eunit
- [x] dialyzer / xref

**Comment by kuenishi** [comment]
No new warning at xref / dialyzer check. https://gist.github.com/kuenishi/2dc060ded55484c68260

File: src/riak_cs_wm_bucket_policy.erl [hunk 1]
```diff
@@ -72,7 +72,7 @@ to_json(RD, Ctx=#context{start_time=_StartTime,
     riak_cs_dtrace:dt_bucket_entry(?MODULE, <<"bucket_get_policy">>,
                                       [], [riak_cs_wm_utils:extract_name(User), Bucket]),
 
-    case riak_cs_s3_policy:bucket_policy(Bucket, RiakPid) of
+    case riak_cs_s3_policy:fetch_bucket_policy(Bucket, RiakPid) of
```
Comment (kuenishi) [inline src/riak_cs_wm_bucket_policy.erl:75]: Note: we should fetch `policy_module` and use it as `PolicyMode:fetch_bucket_policy/2` and so is `response_module` ... and that'd be another issue to refactor.

**Comment by kuenishi** [comment]
Nice refactoring! I also confirmed that there are no more extra `riakc_pb_socket:get/5` on moss.buckets and moss.users. But I also found that bogus get against `riak-cs-gc`, and fetching object manifest twice at one GET. Do you know why?

Despite three new issues found and opened, then +1.

File: src/riak_cs_wm_utils.erl [hunk 2]
```diff
@@ -707,6 +734,17 @@ bucket_owner(Bucket, RiakPid) ->
             undefined
     end.
 
+-spec bucket_owner(riakc_obj:riakc_obj()) -> undefined | acl_owner().
+bucket_owner(BucketObj) ->
```
Comment (reiddraper) [inline src/riak_cs_wm_utils.erl:738]: This doesn't appear to handle the case where `BucketObj` is `notfound`, or do we ensure it's never called with `notfound`?

Comment (reiddraper) [inline src/riak_cs_wm_utils.erl:738]: Ah, looks like there is only one place that calls that function, and it's the function I've been editing in #631.

**Comment by shino** [comment]
Note to self. This change was included in 1.5.0.

**Pull request merged**